[AMD-AIE] Support implicit source/target addressing in amdaie.npu.dma_cpy_nd

Abhishek-Varma opened the pull request:
-- This commit adds a fix for supporting implicit source/target addressing in amdaie.npu.dma_cpy_nd ops as part of `--iree-amdaie-lower-to-aie` pass.

Signed-off-by: Abhishek Varma <abhvarma@amd.com>

Abhishek-Varma commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/lower_to_aie.mlir:
```diff
@@ -300,26 +300,37 @@ module {
-// CHECK-SAME:    @[[OBJ1]]
-// CHECK-NEXT:    aiex.npu.dma_wait
-// CHECK-SAME:    @[[OBJ1]]
+// CHECK-SAME:            %[[ARG0]][1, 1, 0, 32][1, 1, 32, 32][1, 1, 64]
+// CHECK-SAME:            issue_token = true
+// CHECK-SAME:            metadata = @[[OBJ1]]
+// CHECK-NEXT:    aiex.npu.dma_wait {symbol = @[[OBJ1]]}
+// CHECK:         aiex.npu.dma_memcpy_nd
+// CHECK-SAME:            %[[ARG0]][0, 0, 0, 0][0, 0, 32, 64][2048, 2048, 64]
```
Hi @jtuyls - could you check if it looks okay to you?

The former one is for the explicit target addressing whereas the latter for implicit.

Similarly for the two cases below for source addressing.

jtuyls replied:
Yes, this output looks good

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -403,23 +403,79 @@ LogicalResult getStaticDims(Operation *op,
   // Convert bidirectional `amdaie.npu.dma_cpy_nd` op into two halves.
-  if (dmaOp.hasSourceAddressing()) {
+
+  // TODO(avarma): The logic for checking implicit L3 addressing should just be
```
What do you mean with 'checking implicit L3 addressing'?

Abhishek-Varma replied:
So I was trying to refer to the following case :-
```
     %dma_target_l3 = amdaie.circular_dma_cpy_nd(%obj0[] [] [], %obj1[] [] []) : (!amdaie.logicalobjectfifo<memref<32x64xi32>>, !amdaie.logicalobjectfifo<memref<32x32xi32, 1>>)
    controlcode {
        %npu_dma_1 = amdaie.npu.dma_cpy_nd %dma_target_l3([] [] [], [] [] [])
        amdaie.npu.dma_wait(%npu_dma_1, S2MM)
    }
```

Where the access pattern is made implicit in `%npu_dma_1` and I'm trying to see through the `npu.dma_cpy_nd` source if the addressing on L3 is in source or target of `circular_dma_cpy_nd`.

Could you suggest a better way to word the same in comment?

Also, I wanted to confirm if the TODO's intention is valid - as in I wanted to have an API (defined in `AMDAIEOps.td`) within `npu.dma_cpy_nd` that'd answer if the op has implicit/explicit access/addressing patterns. If that's okay, I would like to add that within the same PR and use it in the appropriate place in the fix.

jtuyls replied:
> Also, I wanted to confirm if the TODO's intention is valid - as in I wanted to have an API (defined in AMDAIEOps.td) within npu.dma_cpy_nd that'd answer if the op has implicit/explicit access/addressing patterns. If that's okay, I would like to add that within the same PR and use it in the appropriate place in the fix.

What's the difference between that new api and the existing `hasSourceAddressing`?

jtuyls replied:
> Could you suggest a better way to word the same in comment?

I think the wording 'checking implicit L3 addressing' is ok! I am just not sure about what exact api you're thinking of?

jtuyls replied:
I think `hasImplicitSourceAddressing()` should just be `!hasExplicitSourceAddressing()`. If you also want to check for L3 access, I think it would be better to make that explicit in the function name and imo better to make that a separate function: `hasSourceOnL3` or maybe better `int64_t getSourceMemorySpaceInt()`

Abhishek-Varma replied:
Oh okay, cool.

Then I can add `getSourceMemorySpaceInt` and `getTargetMemorySpaceInt` instead for `npu.dma_cpy_nd` op and make use of those APIs here.

jtuyls replied:
Btw, for consistency, I used `uint64_t getMemorySpaceAsUInt()` before: https://github.com/nod-ai/iree-amd-aie/blob/741facf6ab616529978f077bd75b1a941b7385ac/compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td#L504. And also note that I used `0` for L3

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -403,23 +403,79 @@ LogicalResult getStaticDims(Operation *op,
+    SmallVector<int64_t, 4> staticOffsets;
+    SmallVector<int64_t, 4> staticSizes;
+    SmallVector<int64_t, 3> staticStrides;
+    if (dmaOp.hasSourceAddressing()) {
```
Is this underscore prefix a common pattern for variable naming in MLIR/LLVM codebases? I would avoid it as there are some C++ caveats around naming collisions with prefix underscores.

jtuyls replied:
Also, why are these additional variables with leading underscores needed? Can't we just work with `staticOffsets` etc directly?

Abhishek-Varma replied:
Yes, good point - I shall address that.

jtuyls reviewed (commented):
Not sure whether this is ready for review as you I saw a TODO which I think you want to address in this PR? 

Left a few preliminary comments and I think it would good to add an e2e test as well.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -403,23 +403,79 @@ LogicalResult getStaticDims(Operation *op,
+    staticSizes[sizeIndex--] = shapeArr[i];
+  }
+  int64_t strideIndex = staticStrides.size() - 1;
+  // Since aiex.npu.dma_cpy_nd expects 3-dimensional stride attribute, it has
```
Note the MLIR-AIE bump in this PR which will make the last dimension explicit: https://github.com/nod-ai/iree-amd-aie/pull/503

Abhishek-Varma replied:
Thanks for pointing that out!
I'll accordingly make changes to this revision (and the lit test).

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -401,23 +401,71 @@ LogicalResult getStaticDims(Operation *op,
+  SmallVector<int64_t> shapeArr;
+  for (auto shape : memrefType.getShape()) shapeArr.push_back(shape);
+  int64_t sizeIndex = staticSizes.size() - 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticSizes[sizeIndex--] = shapeArr[i];
+  }
```
You can use `std::copy()` here.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -436,16 +484,28 @@ LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
-                             dmaOp.getTargetMixedStrides(), staticOffsets,
-                             staticSizes, staticStrides))) {
-      return failure();
+    SmallVector<int64_t, 3> staticStrides(4, 1);
```
```suggestion
    SmallVector<int64_t, 4> staticStrides(4, 1);
```

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -401,23 +401,71 @@ LogicalResult getStaticDims(Operation *op,
-                             dmaOp.getSourceMixedStrides(), staticOffsets,
-                             staticSizes, staticStrides))) {
-      return failure();
+    SmallVector<int64_t, 3> staticStrides(4, 1);
```
```suggestion
    SmallVector<int64_t, 4> staticStrides(4, 1);
```

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -436,16 +484,28 @@ LogicalResult npuDmaCpyNdOpToAIE(IRRewriter &rewriter,
   }
-  if (dmaOp.hasTargetAddressing()) {
+  if (dmaOp.hasTargetAddressing() || dmaOp.getTargetMemorySpaceAsUInt() == 0) {
+    // DmaOp has target addressing on L3 either explicitly or implicitly.
```
You're not really ensuring here that the addressing is on L3 side. This condition would still succeed if `dmaOp.hasTargetAddressing() == true` and `dmaOp.getTargetMemorySpaceAsUInt() != 0`, so it might be good to reflect that in the comment.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -401,23 +401,71 @@ LogicalResult getStaticDims(Operation *op,
   rewriter.setInsertionPoint(dmaOp);
   // Convert bidirectional `amdaie.npu.dma_cpy_nd` op into two halves.
-  if (dmaOp.hasSourceAddressing()) {
+  if (dmaOp.hasSourceAddressing() || dmaOp.getSourceMemorySpaceAsUInt() == 0) {
```
You're not really ensuring here that the addressing is on L3 side. This condition would still succeed if `dmaOp.hasSourceAddressing() == true` and `dmaOp.getSourceMemorySpaceAsUInt() != 0`, so it might be good to reflect that in the comment.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -401,23 +401,55 @@ LogicalResult getStaticDims(Operation *op,
+    Operation *op, MemRefType memrefType, SmallVectorImpl<int64_t> &staticSizes,
+    SmallVectorImpl<int64_t> &staticStrides) {
+  // 1. Static sizes.
+  SmallVector<int64_t> shapeArr = llvm::to_vector(memrefType.getShape());
+  std::copy(shapeArr.begin(), shapeArr.end(),
+            staticSizes.end() - shapeArr.size());
```
I think we should check that `staticSizes` is large enough. And it would be good to add a test case for this as well.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:
```diff
@@ -401,23 +401,55 @@ LogicalResult getStaticDims(Operation *op,
+  int64_t strideIndex = staticStrides.size() - 1;
+  // For contiguous access the strides for the last dimension is always 1.
+  int64_t prevStride = 1;
+  for (int64_t i = shapeArr.size() - 1; i >= 0; i--) {
+    staticStrides[strideIndex] = prevStride;
+    prevStride = staticStrides[strideIndex--] * shapeArr[i];
+  }
```
I think we should check sizes here as well to ensure `strideIndex` doesn't become negative? And add a testcase for this.

Abhishek-Varma commented:
Hi @jtuyls - I've addressed your changes.

Note: For the size/stride check only one lit test would suffice because staticSize/staticStride size is same - so we only need to check on shapeArr.size() being greater than that at the beginning of static metadata formation.

jtuyls commented:
> Hi @jtuyls - I've addressed your changes.
> 
> Note: For the size/stride check only one lit test would suffice because staticSize/staticStride size is same - so we only need to check on shapeArr.size() being greater than that at the beginning of static metadata formation.

That's not really guaranteed from the function's perspective right? Someone could change the sizes of the size and stride vectors so they're not the same anymore?

Abhishek-Varma commented:
> > Hi @jtuyls - I've addressed your changes.
> > Note: For the size/stride check only one lit test would suffice because staticSize/staticStride size is same - so we only need to check on shapeArr.size() being greater than that at the beginning of static metadata formation.
> 
> That's not really guaranteed from the function's perspective right? Someone could change the sizes of the size and stride vectors so they're not the same anymore? You could add an assert to ensure the sizes are the same.

Sure. I've added an assert now within the function.
